include/directx/d3dx12.h

```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
     D3D12_NODE_MASK m_Desc;
 };
 
+//------------------------------------------------------------------------------------------------
+// Implementation for the new CheckFeatureSupport API
+
+#ifdef FEATURE_SUPPORT_USE_VARIANT
+#if (__cplusplus >= 201703L || _MSVC_LANG >= 201703L)
+#define FEATURE_SUPPORT_VARIANT_OPT_IN
+#include <variant>
+#else
+static_assert(false, "std::variant is only supported on compilers with C++17 enabled")
+#endif
+#endif
+
+// Macro to set up a getter function for each entry in feature support data
+// The getter function will have the same name as the feature option name
+#define FEATURE_SUPPORT_GET(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION() const \
+{ \
+    return FEATURE.OPTION; \
+}
+
+// Macro to set up a getter function for each entry in feature support data
+// Also specifies the name for the function which can be different from the feature name
+#define FEATURE_SUPPORT_GET_NAME(RETTYPE,FEATURE,OPTION,NAME) \
+RETTYPE NAME() const \
+{\
+    return FEATURE.OPTION; \
+}
+
+// Macro to set up a getter function for feature data indexed by the graphics node ID
+// The default parameter is 0, or the first availabe graphics device node
+#define FEATURE_SUPPORT_GET_NODE_INDEXED(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION(UINT NodeIndex=0) const \
+{\
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+// Macro to set up a getter function for feature data indexed by NodeIndex
+// Allows a custom name for the getter function
+#define FEATURE_SUPPORT_GET_NODE_INDEXED_NAME(RETTYPE,FEATURE,OPTION,NAME) \
+RETTYPE NAME(UINT NodeIndex=0) const \
+{\
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+// Macro to initialize a member feature data struct corresponding to the specified feature
+// Will stop the initialization process and return an error code upon failure
+#define INITIALIZE_MEMBER_DATA_CHECKED(FEATURE, MEMBER) \
+if (FAILED(m_hStatus = m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))) { \
+    return m_hStatus; \
+}
+
+// Macro to initialize a member feature data and report on failure
+// Does not stop the initialization process by itself
+#define INITIALIZE_FAILED(FEATURE, MEMBER) FAILED(m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))
+
+class CD3DX12FeatureSupport
+{
+public:
+    CD3DX12FeatureSupport()
+    : m_pDevice(nullptr)
+    , m_hStatus(E_INVALIDARG)
+    {}
+
+    CD3DX12FeatureSupport(ID3D12Device* pDevice)
+    {
+        Init(pDevice);
+    }
+
+    HRESULT Init(ID3D12Device* pDevice)
+    {

```

Comment:
Continuation from the last PR: Need to figure out how to report the errors to the user in a more intuitive manner.

Right now the status of initialization is stored in m_hStatus. User can use GetStatus() to access this underlying variable, but they must do so explicitly. The factory method returns a variant, so if it will return the error code only if the initialization failed, but that only works on C++ 17 compilers and if the user chooses to turn it on. Also the current factory method design will have different function signatures when the std::variant usage is turned on or off, which is not friendly for the users.

Reply:
Another question is whether we should disable the regular constructor and only allows users to create a CD3D12XFeatureSupport object through the factory method? The regular constructor cannot return anything but the constructed object. It can throw an exception, but not all programmers would want to use exception handling on a constructor.

Reply:
I've decided to keep only the default constructor and the Init() method. See the new comment below.